torchx/c_src/torchx.cpp

```diff
@@ -21,8 +22,89 @@ inline const std::string* type2string(const torch::ScalarType type)
   return nullptr;
 }
 
+// the class instance to manage the refcount of Tensor
+class TensorP
+{
+public:
+  TensorP(ErlNifEnv *env, const ERL_NIF_TERM arg) : ptr(nullptr)
+  {
+    // setup
+    if (!enif_get_resource(env, arg, TENSOR_TYPE, (void **)&ptr))
+    {
+      err = nx::nif::error(env, "Unable to get tensor param in NIF");
+      return;
+    }
+
+    refcount = (std::atomic<int> *)(ptr + 1);
+    deleted = (std::atomic_flag *)(refcount + 1);
+
+    if (refcount->load() == 0)
+    {
+      // already deallocated
+      ptr = nullptr;
```

Comment:
The PR looks good but I am bit confused about this code path that sets the pointer to `nullptr`. This path will only be executed if you try to access a deallocated tensor, so it feels weird that trying to access a deallocated tensor makes pointer changes.

We should either not set the pointer to null or do it on deallocate. I think we don't need to set it to nil because the atomic variables gives us everything we need to know about the tensor state.

Reply:
this ptr is class variable not tensor itself.